### ISSUE TITLE
Hi, can you provide a windows app.

### ISSUE THREAD

**Issue by vanopy** [nagisa/msi-rgb#3]
I need a  windows app because  I don't like to use MSI gaming app, but i want  to control the  rgb light.

Thank you.

**Comment by nagisa** [comment]
Providing an app for Windows would involve writing a kernel driver (as far as I can tell), getting the driver signed (something I have no idea how to do), and figuring out how to make Windows load the driver.

I cannot dedicate time for all the effort needed to release a Windows application myself, but if you, or anybody else, wants to help, I’ll gladly accept it.

**Comment by vanopy** [comment]
Thank you for you reply, i found a  project which can control the msi card without gaming app: https://github.com/Vipeax/MSI-LED-Tool.

Is there any way to make a new tool like this project?

**Comment by nagisa** [comment]
The tool you’ve linked to still uses the libraries from the gaming app to implement the functionality, so it is not truly “without gaming app”.

**Comment by vanopy** [comment]
is there anyway to make a tool with the libraries ?

**Comment by nagisa** [comment]
Sure.

**Comment by stbatuhan** [comment]
So somehow can you use that to manipulate for motherboard on Windows? I tried with ubuntu on windows 10 but not worked of course :( And standart system of MSI for leds is not enough. I purchased 5050 leds and nzxt hue+ set. Hue+ is great but on MSI side not fancy things like that :(

**Comment by nagisa** [comment]
This cannot be used on windows yet. I still haven't seen a standalone ISA
bus driver.

**Comment by stbatuhan** [comment]
Well is there a way to contribute? I'm a programmer also but not interested on hardware programming and drivers on the kernel for now. Also i hate when i get home after work :) Bec all my day passes on the programming. So i spent only almost an hour @last-friday

**Comment by nagisa** [comment]
You could try to write an utility that uses MSI's own drivers, but I would
personally avoid them because of them enabling trivial vulnerabilities.

**Comment by dcerisano** [comment]
RUST, and now particularly RUSTUP seem to have good cross-platform credentials.
I will update this comment with the links.
Peeking and poking at hardware registers is supported by all platforms, so a separate driver may not be necessary. As I see it, msi-rgb is a cross-platform driver.

**Comment by dcerisano** [comment]
Looks like you can run msi-rgb on Windows.

https://msdn.microsoft.com/en-us/commandline/wsl/install_guide

I have not tried it, but this is a full blown user-mode version of Ubuntu from Canonical allowing bash.

Personally I think it is cheating, so I will continue trying to get RUSTUP to compile msi-rgb using the MSVC tool chain.

**Comment by nagisa** [comment]
You cannot as WSL does not give access to ISA ports.

**Comment by dcerisano** [comment]
Have you looked at the x86 crate?

**Comment by dcerisano** [comment]
[This](http://www.highrez.co.uk/downloads/inpout32/) looks to be the missing piece of the puzzle.
There are no crates available for it, but it allows direct user mode ipa port access in Windows.

**Comment by Arlodotexe** [comment]
I just want to point out that I installed the Linux Subsystem for Windows, Ubuntu flavor, and successfully built the project. However, it chokes up with the error: 
```
error: could not open /dev/port; try sudo?
   caused by: No such file or directory (os error 2)
```

I am interested in helping to port this to Windows, but I have hardly worked with hardware development or Windows software development. I've been a JS developer for a few years and am trying to get my foot in the door to building these kinds of apps. If I can be of any help, I will gladly contribute!

**Comment by nagisa** [comment]
The missing piece as pointed by the Dominic seems right. Porting the app
should be just changing the program to load the linked driver and use
functions provided by it on windows rather than the inb and outb functions
called currently.

**Comment by dcerisano** [comment]
Ports are protected in Windows and require signed drivers.
The windows port driver I mentioned is signed.

The interface is standard in/out port protocol.

I have since moved onto usb adressable RGBs,
which do not require leaping through such fiery hoops,

*Dominic Cerisano*
Director Mobile VR Labs
www.standard3d.com <https://standard3d.com>
dcerisano@standard3d.com
ca.linkedin.com/in/dcerisano
dcerisano@skype

**Comment by dcerisano** [comment]
RWeverything is a little more up-to-date and supports both command line and gui.

http://rweverything.com

**Comment by ixjf** [comment]
It seems Mystic Light itself uses WinRing0 for something. The WinRing0.dll is modified and exports some LEDctrl and LEDGetAddress functions. Perhaps something could be done with that?

**Comment by ixjf** [comment]
I have 'ported' msi-rgb to Windows. See [here](https://github.com/ixjf/MSIRGB).

It's not exactly a port. 

For one, it's not written in Rust. I didn't add support for Windows to this project. It was a possibility, and I do find Rust attractive, but 1) Rust is a bit too verbose 2) There are no idiomatic Win32 wrappers and it lacks other little niceties that make C++ far more convenient to get things done quickly 3) I barely have any experience with Rust (that's not to say I didn't try to port it in Rust, I did and I gave up when I realized the amount of verbose code I'd have to write and the amount of time I'd have to spend to write code to make calls to Win32).

For another, while I ported over most of the functions from here, my goal was to do it differently. I.e. reverse engineering Mystic Light to figure out how to do it right, rather than try to guess from trial and error. @nagisa did get it pretty much right - but I couldn't have known.

Also, reverse engineering Mystic Light allowed me to understand a couple of things: Mystic Light-supported MBs use either NCT6797(D), NCT6795(D), NCT5567, or NCT5565. Some have specialized LED controllers, which means they work in a completely different way. Others also work differently though they use these same chips. However, IO to all 4 chips works exactly the same. They're always found at 0x4E. Functionality is not chip-dependent, but motherboard-dependent. So there are going to be motherboards not compatible with msi-rgb even though they use an NCT6795(D) chip. Also it seems to me that LED control on these motherboards affects all built-in MB LED hardware (including headers)? It looked to me like it, but I don't have a motherboard to test it to make sure I'm not missing anything.

**Comment by dcerisano** [comment]
That is excellent work @ixjf.
Did you explore the possibility of using this [signed library](https://github.com/ellysh/InpOut32) that can access ports without admin or kernel privileges?

**Comment by ixjf** [comment]
That is exactly the one I'm using. I just scrapped all the user mode code and wrote my own to avoid a DLL dependency and because I already had code to talk to a driver that was cleaner and simpler (I only needed specific code to load the x64 driver as well). The driver is also embedded in the program and is extracted to the process' current path before loading, and then deleted afterwards on unloading, so it's really all just one 300kb EXE. 

I don't think it's possible to load a kernel driver without admin privileges, though.

**Comment by dcerisano** [comment]
It is not a kernel driver, it is a signed library. Signing authority allows access to protected memory like IO ports. All you really need to do was replace your port register call with the InpOut32 call.

**Comment by ixjf** [comment]
No, it uses a kernel driver. There is a kernel mode and user mode to it: the DLL, which applications load and call functions from, and the kernel mode, where the magic happens. It's exactly because of that kernel driver that it needs to be signed. Only signed drivers can be loaded by Windows. Only drivers can access the hardware at that level.

Also, no, it doesn't. Coincidentally, I was trying to figure that out just now. I had fully turned off my PC and then turned it back on and I realized the LEDs weren't working. So far, nothing.

**Comment by dcerisano** [comment]
The point being (is-a and has-a aside) that it allows user level port access.

To re-initialise your rgb header, just launch MSI GAMING APP and set your leds to white (or any colour). Only have to do this once per cold boot (after unplugged).

**Comment by ixjf** [comment]
Yes, I know, but I want to figure out why it doesn't work at all now. I followed all the main calls in Mystic Light and I don't seem to be missing anything there, so there must be something else initializing the chip.

**Comment by dcerisano** [comment]
It is obviously something that MSI GAMING APP is doing, strongly suspect it is just a sequence of bytes sent to some  bank of port registers on first run.

**Comment by ixjf** [comment]
The DLLs DO matter, because that's where I get the information from. I'm reverse engineering them, not doing trial and error with RWeverything.

Also, no, you can't. Especially not a kernel driver. You need sudo on Linux too, don't you?

**Comment by dcerisano** [comment]
No DLLs don't matter - only the init codes to the SIO. You should be able to get them from tracing the MSI DLLs from cold boot (or RWEverything).

As I said, DLL signing authority is exactly what allows user-level access to protected memory like SIO ports.

**Comment by ixjf** [comment]
I think you don't understand what I'm saying. I'm NOT doing trial and error with RWeverything. That can tell me what is happening, but not how and why. I'm reverse engineering Mystic Light - disassembling the code and seeing what it does.

Also, no. Signing is not what allows user level access to hardware. Signing provides a way for both Windows and the user to know that some executable comes from the author it says it does. That's about it. It's like a real life signature. 
In order to access the hardware, you need to run in kernel space. Hence, you need a kernel driver. The reason kernel drivers must be signed is because those drivers can do pretty much anything, especially without being detected, so it's best only those from reliable sources be installed. Moreover, it's not the InpOut32 DLL that is signed - it's the kernel driver that the DLL loads.

And yes, only an administrator can install a driver. The reason you don't see Mystic Light asking for permissions every time you run it is because when you installed Mystic Light, or the Gaming App, the setup asked you for them. When it installed the program, it installed the service as well and allowed anyone to run it afterwards. If you look at LEDKeeper.exe in Mystic Light's directory, you'll see a small blue and yellow icon on the side that shows that the executable requires admin rights.

**Comment by dcerisano** [comment]
I don't see why you would want to reverse engineer SIO port protocols without the aid of a port sniffer. That seems unnecessarily difficult. It is how @nasiga got this far in the first place. 

Also reverse engineering OEM apps seems extraordinarily laborious given they change the apps every time they release a new SIO chip, mobo design, Tuesday, etc (which is the core point of not doing it).

**Comment by dcerisano** [comment]
But you will have to do it all over again every time MSI sneezes! Not to mention there are whacks of other OEM SIO chips that msi-rgb does NOT handle. 

It is a combinatorial explosion!

**Comment by ixjf** [comment]
Also, it's highly unlikely the program logic is going to change that much from version to version.

**Comment by dcerisano** [comment]
It's not the progam logic, it's the proprietary SIO protocols that are subject to extreme variance from OEM to OEM and even BIOS to BIOS version.

**Comment by ixjf** [comment]
msi-rgb does handle them. That's the thing that nagisa got wrong. As I had already said above, functionality is mostly motherboard-dependent, not chip-dependent. There are probably half of the motherboards that Mystic Light supports that should work with msi-rgb. The other half could be easily implemented as well with this knowledge now.

**Comment by dcerisano** [comment]
Got your leds working yet? Lol.

**Comment by ixjf** [comment]
Mystic Light-supported MBs all use the same 6 chips, which function in exactly the same way. They just have different functionality enabled. There are only a few MBs with different, Renesas chips.

**Comment by dcerisano** [comment]
Case in point - let's see how long it takes you to solve the initialisation problem. That is point where I gave up on this approach.

**Comment by ixjf** [comment]
Yes, if this were to become very popular, there could be. But I doubt most people will ever hear about it, so MSI won't care. Plus, it's not like they're losing any money if someone doesn't use Mystic Light. If anything, it might make more people buy their boards.

**Comment by ixjf** [comment]
Also, testing on all supported mobos is unfair. Did you test msi-rgb on all boards with an NCT6795 chip?

**Comment by dcerisano** [comment]
I would have agreed with you up until now, but apparently some smart guy just opensourced a Windows version of msi-rgb. The space-time continuum has been altered!

**Comment by ixjf** [comment]
Doubt it. I posted it on Reddit and no one batted an eye. Think I'll be fine. Especially if no one makes a GUI for it. N00bs.

**Comment by dcerisano** [comment]
It will takesa while for people to grasp the full extent of this folly. Then you will have youtube tech influencers hounding you for interviews. RGB is a crazed phenomenon!

**Comment by nagisa** [comment]
@ixjf 

> Yes, I know, but I want to figure out why it doesn't work at all now. I followed all the main calls in Mystic Light and I don't seem to be missing anything from there, so there must be something else initializing the chip.

We had the same problem with msi-rgb. [This](https://github.com/nagisa/msi-rgb/blob/master/src/main.rs#L197) is the code that "enables" the advanced mode that makes RGB work. Make sure that you have all steps of initialization in your code as well.

**Comment by nagisa** [comment]
>  As I had already said above, functionality is mostly motherboard-dependent, not chip-dependent. There are probably half of the motherboards that Mystic Light supports that should work with msi-rgb.

This makes me suspect that

1) Either RGB control is not implemented in the SIO (similar to ASUS’ Aura, which possibly has a separate chip on the ISA bus) and either (a) SIO chip is not related to RGB at all; or (b) SIO is passing the configuration to the actual RGB control chip; or (c) something else entirely;
2) SIO chip is separately programmable/configurable (maybe reads its firmware from the flash which happens to store BIOS), but this would only make sense if the configuration would turn-on and turn-off certain functionality, rather than cause fundamental interface differences to occur.

**Comment by nagisa** [comment]
> Also doesn't the Mystic Light EULA contain a non-reverse-engineering clause? The issues could be more than just technical.

Reverse engineering prohibitions are not enforceable in many jurisdictions as reverse engineering is considered fair use. That clause should not be valid anywhere in the US and EU (Disclaimer: IANAL, not legal advice, check with your local lawyer).

**Comment by ixjf** [comment]
I'm pretty sure I just got it. Am gonna test it further but I think I figured out how to initialize the chip properly.

**Comment by ixjf** [comment]
Also, I fixed it by setting a bit I had missed both looking at msi-rgb and Mystic Light. Rip. It works now.

**Comment by nagisa** [comment]
@ixjf That’s https://github.com/nagisa/msi-rgb/issues/6.

**Comment by CalcProgrammer1** [comment]
I'm using inpout32 on Windows for my Asus Aura project and it works great.  I can write directly to SMBus and Super IO registers just fine and has allowed me to do a lot of R/E'ing work without constantly rebooting to Linux.  Feel free to take a look at my implementations for initializing Super IO and scanning for SMBus controllers.  I ported some of the SMBus code from the Linux kernel.

https://gitlab.com/CalcProgrammer1/OpenAuraSDK

**Comment by dcerisano** [comment]
Now we just need someone to port to the hundreds of other OEM SIO RGB controller chips, maintain the resulting frankenstein monster in perpetuity, and we shall realize the opensource dream of RGB ALL THE THINGS.
